File: benchmarks/java/official/src/main/java/McpBench.java
import io.modelcontextprotocol.server.McpServer;
import io.modelcontextprotocol.server.McpSyncServer;
import io.modelcontextprotocol.server.transport.StdioServerTransportProvider;
import io.modelcontextprotocol.json.McpJsonDefaults;
import io.modelcontextprotocol.spec.McpSchema;
import io.modelcontextprotocol.spec.McpSchema.*;

import java.util.List;
import java.util.Map;

public class McpBench {
    public static void main(String[] args) {
        var transport = new StdioServerTransportProvider(McpJsonDefaults.getMapper());

        var helloTool = Tool.builder()
            .name("hello")
            .description("Say hello to someone")
            .inputSchema(new McpSchema.JsonSchema("object",
                Map.of("name", Map.of("type", "string")),
                List.of("name"), null, null, null))
            .build();

        var addTool = Tool.builder()
            .name("add")
            .description("Add two numbers together")
            .inputSchema(new McpSchema.JsonSchema("object",
                Map.of("a", Map.of("type", "number"), "b", Map.of("type", "number")),
                List.of("a", "b"), null, null, null))
            .build();

        McpSyncServer server = McpServer.sync(transport)
            .serverInfo("bench-official", "1.0.0")
            .capabilities(ServerCapabilities.builder().tools(true).build())
            .toolCall(helloTool, (exchange, req) -> {
                var name = req.arguments().get("name").toString();
                return CallToolResult.builder()
                    .addTextContent(String.format("Hello, %s!", name))
                    .build();
            })
            .toolCall(addTool, (exchange, req) -> {
                var a = ((Number) req.arguments().get("a")).doubleValue();
                var b = ((Number) req.arguments().get("b")).doubleValue();
                return CallToolResult.builder()
                    .addTextContent(String.format("{\"sum\":%s}", a + b))
                    .build();
            })
            .build();

        try { Thread.currentThread().join(); } catch (InterruptedException e) {}
    }
}

File: benchmarks/kotlin/official/src/main/kotlin/McpBench.kt
import io.modelcontextprotocol.kotlin.sdk.types.*
import io.modelcontextprotocol.kotlin.sdk.server.Server
import io.modelcontextprotocol.kotlin.sdk.server.ServerOptions
import io.modelcontextprotocol.kotlin.sdk.server.StdioServerTransport
import io.modelcontextprotocol.kotlin.sdk.server.ClientConnection
import kotlinx.coroutines.runBlocking
import kotlinx.io.asSink
import kotlinx.io.asSource
import kotlinx.io.buffered
import kotlinx.serialization.json.*

fun main(): Unit = runBlocking {
    val server = Server(
        Implementation(name = "bench-official", version = "1.0.0"),
        ServerOptions(
            capabilities = ServerCapabilities(tools = ServerCapabilities.Tools(listChanged = true))
        )
    )

    val helloProps = buildJsonObject {
        putJsonObject("name") { put("type", "string") }
    }
    val addProps = buildJsonObject {
        putJsonObject("a") { put("type", "number") }
        putJsonObject("b") { put("type", "number") }
    }

    val helloTool = Tool(
        name = "hello",
        inputSchema = ToolSchema(properties = helloProps, required = listOf("name")),
        description = "Say hello to someone"
    )

    val addTool = Tool(
        name = "add",
        inputSchema = ToolSchema(properties = addProps, required = listOf("a", "b")),
        description = "Add two numbers together"
    )

    val helloHandler: suspend (ClientConnection, CallToolRequest) -> CallToolResult = { _, req ->
        val name = req.arguments!!["name"]?.jsonPrimitive?.content ?: "world"
        CallToolResult(content = listOf(TextContent(text = "Hello, $name!")))
    }

    val addHandler: suspend (ClientConnection, CallToolRequest) -> CallToolResult = { _, req ->
        val a = req.arguments!!["a"]?.jsonPrimitive?.double ?: 0.0
        val b = req.arguments!!["b"]?.jsonPrimitive?.double ?: 0.0
        CallToolResult(content = listOf(TextContent(text = """{"sum":${a + b}}""")))
    }

    server.addTool(helloTool, helloHandler)
    server.addTool(addTool, addHandler)

    val transport = StdioServerTransport(
        System.`in`.asSource().buffered(),
        System.out.asSink().buffered()
    )
    server.createSession(transport)
    // Keep alive until stdin closes
    while (System.`in`.available() >= 0) {
        kotlinx.coroutines.delay(100)
    }
}
